packages/cicero-core/src/util.js

```diff
@@ -0,0 +1,308 @@
+            ergoEngine,
+            name
+        ));
+
+        // Re-compile formulas
+        logicManager.compileLogicSync(true);
+    }
+}
+
+/**
+ * Parses natural language text and if successful returns the contract or clause data
+ * @param {*} parserManager - the parser manager
+ * @param {*} ciceroMarkTransformer - the parser manager
+ * @param {string} input - the text for the clause or contract
+ * @param {string} [currentTime] - the definition of 'now', defaults to current time
+ * @param {number} [utcOffset] - UTC Offset for this execution, defaults to local offset
+ * @param {string} [fileName] - the fileName for the text (optional)
+ * @return {object} the contract or clause data
+ */
+function parseText(parserManager, ciceroMarkTransformer, input, currentTime, utcOffset, fileName) {
+    // Setup
+    const templateMarkTransformer = new TemplateMarkTransformer();
+
+    // Transform text to ciceromark
+    const inputCiceroMark = ciceroMarkTransformer.fromMarkdownCicero(input);
+
+    // Set current time
+    parserManager.setCurrentTime(currentTime, utcOffset);
+
+    // Parse
+    const data = templateMarkTransformer.dataFromCiceroMark({ fileName:fileName, content:inputCiceroMark }, parserManager, {});
+    return data;
+}
+
+/**
+ * add a new state after execution of an operation
+ * @param {object} instance - the contract instance
+ * @param {string} partyName - name of the party that executed the operation
+ * @param {string} operation - name of the operation that was executed
+ * @param {*} output - result of the operation
+ * @param {string} lifecycleState - current state in instance's lifecycle
+ */
+function addHistory(instance, partyName, operation, output, lifecycleState) {
+    const previousHash = instance.history.length !== 0 ? instance.history[instance.history.length-1].currentHash : null;
+    const timestamp = new Date().toISOString();
+    const currentState =  {
+        previousHash: previousHash,
+        partyName: partyName,
+        operation: operation,
+        result: output,
+        timestamp,
+        lifecycleState: lifecycleState
+    };
+    const content = {};
+    content.metadata = instance.metadata;
+    content.history = instance.history;
+    if(instance.parserManager.getTemplate()) {
+        content.grammar = instance.parserManager.getTemplate();
+    }
+    content.models = {};
+    content.scripts = {};
+
+    let modelFiles = instance.getModelManager().getModels();
+    modelFiles.forEach(function (file) {
+        content.models[file.namespace] = file.content;
+    });
+
+    let scriptManager = instance.getScriptManager();
+    let scriptFiles = scriptManager.getScripts();
+    scriptFiles.forEach(function (file) {
+        content.scripts[file.getIdentifier()] = file.contents;
+    });
+
+    content.data = instance.getData();
+
+    content.signatures = instance.contractSignatures;
+
+    const hasher = crypto.createHash('sha256');
+    hasher.update(stringify(content));
```

Comment:
Use template hash

Reply:
Let's move to some shared class/package. TBD.